### PR TITLE
initial example implementation

### PR DESCRIPTION
**Pull request by jackshirazi**
`application` project holds 4 classes:

- main: An interface for an HTTP server
- main: An implementation of the interface using the embedded http server, all requests return the same static HTML page
- main: A basic implementation of the interface using a simple socket-based http server, all requests return the same static HTML page
- main: A client that starts one of the two servers (depending on a property) and executes several requests to the server, then exits

`plugin` project holds 5 classes:

- main: The actual instrumentation of the basic http server in the `application` project
- main: The resource class that tells the Elastic Java Agent to load that instrumentation class
- test: A MockApmServer which accepts comms from the Java agent and always returns an empty JSON {} struct. If the incoming comm is a transaction, it will hold the JSON for that transaction in an array that can be accessed
- test: An abstract instrumentation test class that sets up a test with a mock APM server and an embedded elastic agent
- test: The actual instrumentation test class which exercises the instrumentation implemented in this project

### CODE REVIEW

**Comment by jackshirazi** [comment]
Note the following is already scheduled work for after this PR is approved:

- add an attribute to plugin instrumentation and regression test
- add tests for the example plugin application 
- flesh out comments and instructions in the plugin README

File: plugin/src/main/java/co/elastic/apm/example/webserver/plugin/ExampleHttpServerInstrumentation.java
Comment (eyalkoren) [inline plugin/src/main/java/co/elastic/apm/example/webserver/plugin/ExampleHttpServerInstrumentation.java]: Add something about the returned scope, which is used at the exit advice as `@Advice.Enter`. Worth mentioning this is a best-practice for scope handling

File: plugin/src/main/java/co/elastic/apm/example/webserver/plugin/ExampleHttpServerInstrumentation.java
Comment (eyalkoren) [inline plugin/src/main/java/co/elastic/apm/example/webserver/plugin/ExampleHttpServerInstrumentation.java:225]: For good measures, wrap in `try` and do `span.end()` and `scope.close()` in a `finally` block as a best-practice example for plugin developers.

File: plugin/src/test/java/co/elastic/apm/mock/MockApmServer.java
Comment (eyalkoren) [inline plugin/src/test/java/co/elastic/apm/mock/MockApmServer.java:63]: Consider throwing `TimeoutException`

File: plugin/src/test/java/co/elastic/apm/mock/MockApmServer.java
Comment (eyalkoren) [inline plugin/src/test/java/co/elastic/apm/mock/MockApmServer.java:98]: This side effect should be stated within the javadoc of this method.
Personally, I think an API that only gets transactions (all or by index) and a separate one for reset is more flexible and clear.

File: plugin/src/test/java/co/elastic/apm/mock/MockApmServer.java
Comment (eyalkoren) [inline plugin/src/test/java/co/elastic/apm/mock/MockApmServer.java]: Doesn't really matter in this mock reporter, only as a style suggestion:
I am not sure why you want to allow other threads to concurrently try to invoke this method, but if that's the intention, you can use `transactions.wait()`, which releases the lock, as opposed to `Thread.sleep()` and wrap the entire block with `synchronized(transactions)`.
This is also useful because it will allow you to `wait` until the timeout instead of busy-sleep, relying on `notify()` when a transaction is received and added.

File: plugin/src/test/java/co/elastic/apm/mock/MockApmServer.java
Comment (eyalkoren) [inline plugin/src/test/java/co/elastic/apm/mock/MockApmServer.java]: Did you keep that intentionally?

**Review by eyalkoren — APPROVED** [review]
Very nice standalone project! 👏 
Since this is only an example, feel free to ignore anything that you disagree with or don't want to spend time on changing. I could

One do think worth a bit more time is the documentation of `ExampleHttpServerInstrumentation`, which is actually the main focus here. Why isn't it commented as javadoc? That would improve how links, formats and lists are rendered. 
Also, I suggest you go over this instrumentation one more time and add a comment for anything you recall you had to think about (like returned scope used in exit advice).

File: plugin/src/test/java/co/elastic/apm/mock/MockApmServer.java
Comment (jackshirazi) [reply plugin/src/test/java/co/elastic/apm/mock/MockApmServer.java]: No, thanks removed now

File: plugin/src/test/java/co/elastic/apm/mock/MockApmServer.java
Comment (jackshirazi) [reply plugin/src/test/java/co/elastic/apm/mock/MockApmServer.java:63]: yes, much better, thanks

File: plugin/src/test/java/co/elastic/apm/mock/MockApmServer.java
Comment (jackshirazi) [reply plugin/src/test/java/co/elastic/apm/mock/MockApmServer.java:98]: If this was a generic class we are providing as a dependency, I agree. As it's currently an example, I'll leave just the one method, but have renamed it and added the doc change, thanks! We should review whether we want to make this a supported class somewhere

File: plugin/src/test/java/co/elastic/apm/mock/MockApmServer.java
Comment (jackshirazi) [reply plugin/src/test/java/co/elastic/apm/mock/MockApmServer.java]: You are absolutely right! I started with a CopyOnWriteArrayList which needed the busy loop, then switched to synchronized ArrayList and forgot to switch to wait-notify lol. The concurrent capability more about the server being updated in a separate thread than it is accessed, but it's also feasible that a test could access concurrently ... though not with the current examples. Fixed now

File: plugin/src/main/java/co/elastic/apm/example/webserver/plugin/ExampleHttpServerInstrumentation.java
Comment (jackshirazi) [reply plugin/src/main/java/co/elastic/apm/example/webserver/plugin/ExampleHttpServerInstrumentation.java:225]: Do you mean like 

```
            Span span = Span.current();
            try {
                if (thrown != null) {
                    span.setStatus(StatusCode.ERROR);
                    span.recordException(thrown);
                }
            } finally {
                span.end();
                Scope scope = (Scope) scopeObject;
                scope.close();
            }
```

Why this is better practice, what could fail to be executed in the original code? Is it the span.end() that could throw an exception?

File: plugin/src/main/java/co/elastic/apm/example/webserver/plugin/ExampleHttpServerInstrumentation.java
Comment (jackshirazi) [reply plugin/src/main/java/co/elastic/apm/example/webserver/plugin/ExampleHttpServerInstrumentation.java]: Yes, the documentation needs a full overhaul as you say. I have no idea why I used comments instead of javadocs! Fixed now

File: plugin/src/main/java/co/elastic/apm/example/webserver/plugin/ExampleHttpServerInstrumentation.java
Comment (jackshirazi) [reply plugin/src/main/java/co/elastic/apm/example/webserver/plugin/ExampleHttpServerInstrumentation.java:225]: I could do this, but it looks odd (scope variable not used)

```
            Span span = Span.current();
            try (Scope scope = (Scope) scopeObject) {
                span.end();
            } finally {
                if (thrown != null) {
                    span.setStatus(StatusCode.ERROR);
                    span.recordException(thrown);
                }
            }
```

Comment (jackshirazi) [reply plugin/src/main/java/co/elastic/apm/example/webserver/plugin/ExampleHttpServerInstrumentation.java:225]: I'm going to merge the current implementation, but have opened a [new issue](https://github.com/elastic/apm-agent-java-plugin-example/issues/2) to discuss this

Comment (eyalkoren) [reply plugin/src/main/java/co/elastic/apm/example/webserver/plugin/ExampleHttpServerInstrumentation.java:225]: I would even go as far as:
```
        try {
            Span span = Span.current();
            try {
                if (thrown != null) {
                    span.setStatus(StatusCode.ERROR);
                    span.recordException(thrown);
                }
            } finally {
                span.end();
            }
        } finally {
            Scope scope = (Scope) scopeObject;
            scope.close();
        }
```

> what could fail to be executed in the original code?

Everything - any silly agent bug in the implementation of any of those can cause an error. Follow the `span.recordException(thrown)` route and imagine there's a wrong assumption for something never to be `null`, overlooking some OTel scenarios, causing this error. You would rightfully claim it is on us to fix, but it is a better practice nevertheless to make sure that spans are ended (and reported - lower impact) and do not leak the thread context (which can cause issues that are hard to debug).

But even more importantly, this is an example for plugin authors, where we can expect not only ours, but their bugs as well. If you guard this way, you restrict the problems to your plugin spans without leaking them to other agent instrumentations.

**Pull request merged**
